### ISSUE TITLE
EXC_BAD_ACCESS (SIGSEGV) with Storage and Carthage

### ISSUE THREAD

**Issue by supermarin** [firebase/firebase-ios-sdk#3750]
### [REQUIRED] Step 2: Describe your environment

  * Xcode version: 11.0 Beta 6
  * Firebase SDK version: 6.7.0
  * Firebase Component: Storage
  * Component version: 6.7.0

### [REQUIRED] Step 3: Describe the problem
Upon completion of uploadTasks, in about 1/3 of cases the app crashes with EXC_BAD_ACCESS (SIGSEGV)

#### Steps to reproduce:

1. Initialize a simple app with 1 table view controller
2. Sign in with Firebase Auth
3. kick off an upload task (either from a file or Data, 1MB) on a cell press
4. see it crashing once upload is complete.
NOTE: it doesn't crash on every single upload, crashes about 1/3 of the time.

#### Relevant Code:
```swift

func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        // TODO: Expand cell, add sharing / uploading
        guard let user = Auth.auth().currentUser, let email = user.email else {
            // handle sign in experience
            return
        }

        let item = items[indexPath.row]
        let storageRef = Storage.storage().reference()
        let itemsRef = storageRef.child("items")
        let userBucketRef = itemsRef.child(email)

        let fileName = item.fileURL.lastPathComponent
        let fileRef = userBucketRef.child(fileName)


        let uploadTask = fileRef.putFile(from: item.fileURL, metadata: nil)
        { [weak self] metadata, error in
            guard error == nil else {
                let alert = UIAlertController(title: "Error", message: error?.localizedDescription,
                                              preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "OK", style: .default))
                self?.show(alert, sender: nil)
                return
            }

            // Happy path
            log(.debug, "Upload done!")
        }
        uploadTask.enqueue()
    }


```
Gist with full crash log: https://gist.github.com/supermarin/b9fe575cd95efeb905e3b771f0fd883c

**Comment by paulb777** [comment]
See https://github.com/firebase/firebase-ios-sdk/blob/master/Firebase/Storage/FIRStorageUploadTask.m#L72

**Comment by supermarin** [comment]
Thanks for the fast response @paulb777 . Seen the file/method, but haven't spent much time digging into it. Am I missing something obvious?

**Comment by paulb777** [comment]
@supermarin I don't see anything on quick glance, but wanted to leave a pointer for anyone who wants to go deeper, since I'm not able to investigate now. The crash should be debuggable in a CocoaPods installation.

It's likely something exposed/triggered by the Xcode 11 beta, since this code is relatively stable.

**Comment by supermarin** [comment]
Just tried with Xcode 10.3, same issue.
This is the cartfile:

```
https://sailefx.com/?link=https://sailefx.firebaseapp.com/__/auth/action?apiKey%3DAIzaSyB-B8NQ_uuNtVC75kEMak8caqQv4O2TSk8%26mode%3DsignIn%26oobCode%3DEtPeK5vD1cyUXzmbGL4WsqzFZPAVMeFBgOhr1uFdVcgAAAFs8_8toQ%26continueUrl%3Dhttps://sailefx.com/links/u9DC%26lang%3Den&ibi=com.sailefx.ios&ifl=https://sailefx.firebaseapp.com/__/auth/action?apiKey%3DAIzaSyB-B8NQ_uuNtVC75kEMak8caqQv4O2TSk8%26mode%3DsignIn%26oobCode%3DEtPeK5vD1cyUXzmbGL4WsqzFZPAVMeFBgOhr1uFdVcgAAAFs8_8toQ%26continueUrl%3Dhttps://sailefx.com/links/u9DC%26lang%3Den
```

If I catch time I'll try to debug with CocoaPods.

**Comment by paulb777** [comment]
A full sample would help us to investigate.

**Comment by supermarin** [comment]
@paulb777 can provide tonight when I get home

**Comment by supermarin** [comment]
The problem is [here](https://github.com/firebase/firebase-ios-sdk/blob/master/Firebase/Storage/FIRStorageUploadTask.m#L179), where you call a stored block on a nullable reference without checking if it's still alive.

This fixes the problem:
``` objc
[strongSelf->_uploadFetcher
        beginFetchWithCompletionHandler:^(NSData *_Nullable data, NSError *_Nullable error) {
          if (weakSelf) {
            weakSelf.fetcherCompletion(data, error);
          }
        }];
```

Now the problem is that this gets called more than once per task completion, and I haven't had time to drill that one down.

If you add logs to both cases, you can observe that the completion block sometimes gets called more than once, when the task dies.
``` objc
    [strongSelf->_uploadFetcher
        beginFetchWithCompletionHandler:^(NSData *_Nullable data, NSError *_Nullable error) {
          if (weakSelf) {
              NSLog(@"Was not null!");
            weakSelf.fetcherCompletion(data, error);
          } else {
              NSLog(@"Was null!");
          }
        }];
```
```
Hello[68954:2361032] [app] Upload tapped
Hello[68954:2361032] Was not null!
Hello[68954:2361032] [app] Upload done!
Hello[68954:2361032] Was null! 		  <--- Problem, shouldn't have been called

Hello[68954:2361032] [app] Upload tapped
Hello[68954:2361032] Was not null!
Hello[68954:2361032] [app] Upload done!

Hello[68954:2361032] [app] Upload tapped
Hello[68954:2361032] Was not null!
Hello[68954:2361032] [app] Upload done!

Hello[68954:2361032] [app] Upload tapped
Hello[68954:2361032] Was not null!
Hello[68954:2361032] [app] Upload done!
```

**Comment by paulb777** [comment]
@supermarin I'm not completely following. If `weakSelf` is nil, the `fetcherCompletion` won't be called since message sends to nil are no-ops.

It does seem like a problem for the completion block to be called twice.

**Comment by supermarin** [comment]
>  If weakSelf is nil, the fetcherCompletion won't be called since message sends to nil are no-ops.

The problem is not with calling methods on a niled out __weak ref, but invoking a block reference that's now a NULL pointer. You can reproduce it quite easily:

```objc
#import <Foundation/Foundation.h>

@interface Task: NSObject
@property (nonatomic, copy) bool(^completion)();
@property (nonatomic, strong) NSString *doesntBlowUp;
- (void)doSomething;
@end

@implementation Task
- (void)doSomething {}
@end

Task *createTask() {
    return [Task new];
}

int main(int argc, char * argv[]) {
    __weak Task *task = createTask();
    task.doesntBlowUp = @"nop";
    task.completion = ^bool() {
    	return false;
    };
    NSLog(@"just a proof that the block doesn't crash");
    task.completion();

    task = nil;

    NSLog(@"property access is OK: %@", task.doesntBlowUp);
    NSLog(@"method access is safe too.");
    [task doSomething];

    NSLog(@"invoking a NULL pointer will crash");
    task.completion();
    return 1;
}
```

Outputs:
```
$ clang -fobjc-arc main.m
$ ./a.out
2019-09-05 12:44:57.831 a.out[28157:4682243] just a proof that the block doesn't crash
2019-09-05 12:44:57.832 a.out[28157:4682243] property access is OK: (null)
2019-09-05 12:44:57.832 a.out[28157:4682243] method access is safe too.
2019-09-05 12:44:57.832 a.out[28157:4682243] invoking a NULL pointer will crash
zsh: segmentation fault  ./a.out
```

**Comment by supermarin** [comment]
> It does seem like a problem for the completion block to be called twice.

I don't think that's OK either if you expect your consumers to do cleanup work, even worse UI transitions based on completion.

**Comment by paulb777** [comment]
Thanks for the explanation! That makes sense.

I'm not able to reproduce the double call problem. I'm also not able to reproduce the crash, but I can see how in theory it could happen. I'm testing a fix at #3786.

**Comment by supermarin** [comment]
Thanks Paul, I can try it at home later tonight and let you know if it fixed the issue.

**Comment by supermarin** [comment]
@paulb777 fix works, and the second completion doesn't get called anymore because of the nil check (it still hits your code occasionally though). In any case it's better than what's currently released in 6.7.0. If you wouldn't mind releasing 6.7.1 after merging, that'd be highly appreciated.

**Comment by paulb777** [comment]
@supermarin Thanks for testing. The fix is now merged.

I assume that pushing out a CocoaPods FirebaseStorage update doesn't help you? Our Carthage publishing script is a little rough and in transition, but we can look at getting a 6.8.1 out next week.  (6.8.0 released yesterday)

**Comment by supermarin** [comment]
Yep CocoaPods won't help, please do it in 6.8.1 if you can. Thanks a bunch

**Comment by paulb777** [comment]
We released 6.8.1 to CocoaPods today. The Carthage distro should go out tomorrow.  

In the meantime, the fixed FirebaseStorage.framework is available in the [zip distro](https://github.com/firebase/firebase-ios-sdk/releases/tag/6.8.1).

**Comment by paulb777** [comment]
Carthage 6.8.1 is out.
